From: Jiri Slaby <jslaby@suse.cz>
Subject: x86: nospec on microcode blacklist
Patch-mainline: Never, SUSE specific
References: bsc#1068032 CVE-2017-5754

In patch patches.suse/21-x86-spec-add-nospec-chicken-bit.patch, we add
"nospec" kernel parameter to disable spectre. Spectre fixes in 4.4.144
add blacklisting based on microcode.  Make sure that we disables also
the former by our nospec().

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 arch/x86/include/asm/spec_ctrl.h |    1 +
 arch/x86/kernel/cpu/intel.c      |    1 +
 arch/x86/kernel/cpu/spec_ctrl.c  |    2 +-
 3 files changed, 3 insertions(+), 1 deletion(-)

--- a/arch/x86/include/asm/spec_ctrl.h
+++ b/arch/x86/include/asm/spec_ctrl.h
@@ -93,5 +93,6 @@ void x86_disable_ibrs(void);
 unsigned int x86_ibrs_enabled(void);
 unsigned int x86_ibpb_enabled(void);
 void x86_spec_check(void);
+int nospec(char *str);
 #endif /* __ASSEMBLY__ */
 #endif /* _ASM_X86_SPEC_CTRL_H */
--- a/arch/x86/kernel/cpu/intel.c
+++ b/arch/x86/kernel/cpu/intel.c
@@ -200,6 +200,7 @@ static void early_init_intel(struct cpui
 		setup_clear_cpu_cap(X86_FEATURE_INTEL_STIBP);
 		setup_clear_cpu_cap(X86_FEATURE_SSBD);
 		setup_clear_cpu_cap(X86_FEATURE_SPEC_CTRL_SSBD);
+		nospec("");
 	}
 
 	/*
--- a/arch/x86/kernel/cpu/spec_ctrl.c
+++ b/arch/x86/kernel/cpu/spec_ctrl.c
@@ -79,7 +79,7 @@ void x86_spec_check(void)
 }
 EXPORT_SYMBOL_GPL(x86_spec_check);
 
-static int __init nospec(char *str)
+int nospec(char *str)
 {
 	setup_clear_cpu_cap(X86_FEATURE_SPEC_CTRL);
 	ibrs_state = 0;
